pkg/services/govmomi/vcenter/clone.go

```diff
@@ -390,6 +405,126 @@ func getDiskConfigSpec(disk *types.VirtualDisk, diskCloneCapacityKB int64) (type
+					ThinProvisioned: types.NewBool(true),
+					VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
+						FileName: "",
+					},
+				},
+				ControllerKey: controller.GetVirtualController().Key,
+			},
+			CapacityInKB: int64(dataDisk.SizeGiB) * 1024 * 1024,
+		}
+
+		// Assign unit number to the next slot on the controller.
+		if err := assignUnitNumber(ctx, dev, devices, additionalDisks, controller, unitOffset); err != nil {
+			return nil, errors.Wrap(err, "failed to assign device unit number to disk")
+		}
+
+		// Update unitOffset to current device unitNumber as starting point for next disk to be assigned.
+		unitOffset = *dev.UnitNumber
+
+		diskConfigSpec := types.VirtualDeviceConfigSpec{
+			Device:        dev,
+			Operation:     types.VirtualDeviceConfigSpecOperationAdd,
+			FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
+		}
+
+		log.V(4).Info("Generated device", "dev", dev)
+
+		additionalDisks = append(additionalDisks, &diskConfigSpec)
+	}
+
+	return additionalDisks, nil
+}
+
+// assignUnitNumber assigns a controller unit number to a device.
+func assignUnitNumber(ctx context.Context, device types.BaseVirtualDevice, existingDevices object.VirtualDeviceList, newDevices []types.BaseVirtualDeviceConfigSpec, controller types.BaseVirtualController, offset int32) error {
+	if offset > maxUnitNumber {
+		return errors.Errorf("%d exceeds maximum number of units %d", offset, maxUnitNumber)
+	}
+
+	if controller == nil {
+		return errors.New("controller parameter cannot be nil")
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	vd := device.GetVirtualDevice()
+	vd.ControllerKey = controller.GetVirtualController().Key
+	vd.UnitNumber = &offset
+
+	units := make([]bool, maxUnitNumber)
+
+	for i := 0; i < int(offset); i++ {
+		units[i] = true
+	}
+
+	sc, ok := controller.(types.BaseVirtualSCSIController)
+	if ok {
+		//  The SCSI controller sits on its own bus
+		log.V(4).Info(fmt.Sprintf("Marking SCSI Controller's unit number: %d", sc.GetVirtualSCSIController().ScsiCtlrUnitNumber))
+		units[sc.GetVirtualSCSIController().ScsiCtlrUnitNumber] = true
+	}
+
+	key := controller.GetVirtualController().Key
+
+	// Check all existing devices
+	for _, device := range existingDevices {
+		d := device.GetVirtualDevice()
+		if d.ControllerKey == key && d.UnitNumber != nil {
+			units[int(*d.UnitNumber)] = true
+		}
+	}
+
+	// Check new devices
+	for _, device := range newDevices {
+		d := device.GetVirtualDeviceConfigSpec().Device.GetVirtualDevice()
+		if d.ControllerKey == key && d.UnitNumber != nil {
+			units[int(*d.UnitNumber)] = true
+		}
+	}
+
+	// Assign first unused unit number
+	for unit, used := range units {
+		if !used {
+			unit32 := int32(unit)
+			vd.UnitNumber = &unit32
+			log.V(4).Info(fmt.Sprintf("Determined next available unit number: %d", unit32))
+			return nil
+		}
+	}
+
+	// If we are here, we did not find a unit number.  Return error.
+	return errors.Errorf("all unit numbers are already in-use")
+}
```

Comment:
Sorry for the big refactor, but I think this simplifies it a bit.

```suggestion
func createDataDisks(ctx context.Context, dataDiskDefs []infrav1.VSphereDisk, devices object.VirtualDeviceList) ([]types.BaseVirtualDeviceConfigSpec, error) {
	log := ctrl.LoggerFrom(ctx)
	additionalDisks := []types.BaseVirtualDeviceConfigSpec{}

	disks := devices.SelectByType((*types.VirtualDisk)(nil))
	if len(disks) == 0 {
		return nil, errors.Errorf("Invalid disk count: %d", len(disks))
	}

	// There is at least one disk
	primaryDisk := disks[0].(*types.VirtualDisk)

	// Get the controller of the primary disk.
	controller, ok := devices.FindByKey(primaryDisk.ControllerKey).(types.BaseVirtualController)
	if !ok {
		return nil, errors.Errorf("unable to find controller with key=%v", primaryDisk.ControllerKey)
	}
	controllerKey := controller.GetVirtualController().Key

	unitNumberAssigner, err := newUnitNumberAssigner(controller, devices, additionalDisks)
	if err != nil {
		return nil, err
	}

	for i, dataDisk := range dataDiskDefs {
		dev := &types.VirtualDisk{
			VirtualDevice: types.VirtualDevice{
				Key: devices.NewKey() - int32(i),
				Backing: &types.VirtualDiskFlatVer2BackingInfo{
					DiskMode:        string(types.VirtualDiskModePersistent),
					ThinProvisioned: types.NewBool(true),
					VirtualDeviceFileBackingInfo: types.VirtualDeviceFileBackingInfo{
						FileName: "",
					},
				},
				ControllerKey: controller.GetVirtualController().Key,
			},
			CapacityInKB: int64(dataDisk.SizeGiB) * 1024 * 1024,
		}

		vd := dev.GetVirtualDevice()
		vd.ControllerKey = controllerKey
		unitNumber, err := unitNumberAssigner.assign()
		if err != nil {
			return nil, err
		}
		vd.UnitNumber = &unitNumber

		log.V(4).Info("Created device for data disk device", "name", dataDisk.Name, "spec", dataDisk, "device", dev)

		additionalDisks = append(additionalDisks, &types.VirtualDeviceConfigSpec{
			Device:        dev,
			Operation:     types.VirtualDeviceConfigSpecOperationAdd,
			FileOperation: types.VirtualDeviceConfigSpecFileOperationCreate,
		})
	}

	return additionalDisks, nil
}

type unitNumberAssigner struct {
	used   []bool
	offset int32
}

func newUnitNumberAssigner(controller types.BaseVirtualController, existingDevices object.VirtualDeviceList, newDevices []types.BaseVirtualDeviceConfigSpec) (*unitNumberAssigner, error) {
	if controller == nil {
		return nil, errors.New("controller parameter cannot be nil")
	}

	used := make([]bool, maxUnitNumber)

	// SCSIControllers also use a unit.
	if scsiController, ok := controller.(types.BaseVirtualSCSIController); ok {
		used[scsiController.GetVirtualSCSIController().ScsiCtlrUnitNumber] = true
	}

	controllerKey := controller.GetVirtualController().Key

	// Mark all unit numbers of existing devices as used
	for _, device := range existingDevices {
		d := device.GetVirtualDevice()
		if d.ControllerKey == controllerKey && d.UnitNumber != nil {
			used[*d.UnitNumber] = true
		}
	}

	// Mark all unit numbers of new devices as used
	for _, device := range newDevices {
		d := device.GetVirtualDeviceConfigSpec().Device.GetVirtualDevice()
		if d.ControllerKey == controllerKey && d.UnitNumber != nil {
			used[*d.UnitNumber] = true
		}
	}

	// Set offset to 0, it will auto-increment on the first assignment.
	return &unitNumberAssigner{used: used, offset: 0}, nil
}

func (a *unitNumberAssigner) assign() (int32, error) {
	if int(a.offset) > len(a.used) {
		return -1, fmt.Errorf("all unit numbers are already in-use")
	}

	for i, isInUse := range a.used[a.offset:] {
		unit := int32(i) + a.offset
		if !isInUse {
			a.used[unit] = true
			a.offset++
			return unit, nil
		}
	}
	return -1, fmt.Errorf("all unit numbers are already in-use")
}
```